Fix #3292: Add check for test files presence for prod files

Sparsh1212 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #3292: Add check for test files presence for prod files

The script traverses different layers of the codebase to check for the presence of test files for prod files.

**Note:** It works on the logic of searching Filename+'Test'.kt in the codebase, to ensure the test file presence.
The files listed in the `test_file_exemptions.textproto` will be skipped for this check, and currently, all the prod files which were failing this check, have been added to this list so that the workflow doesn't fail. 
We can remove any file from the exemption list, where we think that we can't go without a test file for that prod file, or if a test file is now added for that particular file.

To run the script, use:

`bazel run //scripts:test_file_check -- $(pwd)`

For testing the script, automated tests have been added.
To execute the tests, use:

`bazel test //scripts/src/javatests/org/oppia/android/scripts/testfile:TestFileCheckTest`

**Note:** We are generating the test assets dynamically at the time of executing them. The test assets are automatically deleted, when the test finishes.

**Blocked on:** #3341 #3340

Screenshot of all new added tests passing successfully on local:

![image](https://user-images.githubusercontent.com/55937724/123287748-99c47b00-d52c-11eb-8c60-4bf1a1f53050.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sparsh1212 committed:
Fix #3290: Add support for generic regex pattern matching

Sparsh1212 committed:
Apply review suggestions for proto location

Sparsh1212 committed:
Delete main directory

Sparsh1212 committed:
Implment script logic and tests

Sparsh1212 committed:
CI setup part 1

Sparsh1212 committed:
syntax fix in yaml file

Sparsh1212 committed:
import nits in dummy file

Sparsh1212 committed:
Use regex patterns from script in script test

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Improve naming of script_assets variables

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Revamp testing approach

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Fix #3291: Add check for XML syntax correctness

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Introduce XML syntax check in static checks workflow

Sparsh1212 committed:
Fix #3292: Add check for test files presence for prod files

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -0,0 +1,613 @@
+package org.oppia.android.scripts
+
+class ExemptionsList {
```
do we need a class or could we use `object`?

Sparsh1212 replied:
Okay!
Changed it to an object

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,88 @@
+      return validPaths
+    }
+
+    fun checkIfAllowedDirectory(
```
add kdocs

Sparsh1212 replied:
Done

anandwana001 reviewed (changes requested):
Thanks, @Sparsh1212 

The same comment applies here from your previous PR regarding the script `.kt` file and its test file.

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 commented:
> Thanks, @Sparsh1212
> 
> The same comment applies here from your previous PR regarding the script `.kt` file and its test file.

Thanks, @anandwana001 for the suggestions.
The PR is now updated as per the suggestions. Please take a look.

fsharpasharp reviewed (approved):
Approving for BUILD.bazel change

Sparsh1212 committed:
Add ScriptConstants file to exemptions test list

Sparsh1212 committed:
Add if: always()

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Add if: always()

anandwana001 commented:
@Sparsh1212 Please follow in all your PRs 
https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section

Sparsh1212 committed:
Apply review suggestions on PR

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make review suggestions from regex pattern checks

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Implement review suggestions based on #3340

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -0,0 +1,614 @@
+package org.oppia.android.scripts
+
+object ExemptionsList {
```
add kdocs

Sparsh1212 replied:
done

BenHenning replied:
Perhaps: ``ScriptExemptions``?

Sparsh1212 replied:
SGTM. Done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -0,0 +1,614 @@
+    "NetworkModule.kt",
+    "ExemptionsList.kt",
+    "ScriptResultConstants.kt",
+    )
+}
```
```suggestion
  )
}
```

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,132 @@
+
+class TestFileCheck {
+  companion object {
+
```
no empty line

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,132 @@
+import org.oppia.android.scripts.ExemptionsList
+import org.oppia.android.scripts.ScriptResultConstants
+
+class TestFileCheck {
```
add kdocs

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,132 @@
+       * @param allowedDirectories a list of all the directories which needs to be checked
+       * @return all files which needs to be checked
+       */
+      fun collectSearchFiles(
+        repoPath: String,
+        allowedDirectories: List<String>, exemptionsList: Array<String>)
+        : List<File> {
+        return File(repoPath).walk().filter { it ->
+          checkIfAllowedDirectory(
+            retrieveFilePath(it, repoPath),
+            allowedDirectories)
+            && it.isFile
+            && it.name.endsWith(".kt")
+            && it.name !in exemptionsList
+        }.toList()
+      }
```
```suggestion
      fun collectSearchFiles(
        repoPath: String,
        allowedDirectories: List<String>, exemptionsList: Array<String>
      )
        : List<File> {
        return File(repoPath).walk().filter { it ->
          checkIfAllowedDirectory(
            retrieveFilePath(it, repoPath),
            allowedDirectories
          )
            && it.isFile
            && it.name.endsWith(".kt")
            && it.name !in exemptionsList
        }.toList()
      }
```

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,132 @@
+     * @param prodFileName name of the prod file
+     * @return potential name of the test file
+     */
+    fun retrievePotentionalTestFileName(prodFileName: String): String {
```
can be private function

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,132 @@
+     * @param matchedFileNames list of all the files missing a test file
+     * @return log the failures
+     */
+    fun logFailures(matchedFileNames: List<String>) {
```
can be private function

Sparsh1212 replied:
done

anandwana001 reviewed (changes requested):
Suggesting to follow the same suggestion PR review comments from #3340 pr

Sparsh1212 committed:
Refactor PR as per feedback recieved

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement feedback suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 commented:
> Suggesting to follow the same suggestion PR review comments from #3340 pr

Thanks, @anandwana001. 
The PR is updated on the basis of comments received on #3340 and new suggestions too. PTAL.

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

anandwana001 commented:
I will take a look at it tomorrow morning.

Sparsh1212 committed:
nits

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

anandwana001 reviewed (changes requested):
Please check on failing test

Sparsh1212 committed:
nit fix

Sparsh1212 commented:
> Please check on failing test

These are due to the same issue as discussed [here](https://github.com/oppia/oppia-android/pull/3341#issuecomment-869203395).

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

vinitamurthi reviewed (approved):
LGTM for codeowner files

anandwana001 reviewed (approved):
LGTM

anandwana001 commented:
@Sparsh1212 If you want Ben's review on this PR, please follow the Assignee and Reviewer Guide

https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

BenHenning commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -0,0 +1,617 @@
+    "ExemptionsList.kt",
+    "ScriptResultConstants.kt",
+    "RepoFile.kt",
+    "SyntaxErrorHandler.kt"
```
Please make sure all of the scripts classes have tests--we shouldn't add new files here often.

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -0,0 +1,617 @@
+
+/** Lists of files which should be exepmted for the different script checks. */
+object ExemptionsList {
+  /** List of files which should be exempted for the test file check. */
```
```suggestion
  /**
   * List of files which should be exempted for the test file check.
   *
   * NOTE TO DEVELOPERS: New files should rarely be added here. In general, all new new files should have tests. If you think there's a valid exception, please specifically call this out in your PR description.
   */
```

The notice is important to inform developers when to change this list. The suggestion may need line wrapping.

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -0,0 +1,617 @@
+package org.oppia.android.scripts
```
Please make sure this class has a special codeowners at the end of the CODEOWNERS file & add me as the codeowner. Since this is an exemption list we should make sure that the codeowners for this file are making sure we don't extend the list unless absolutely necessary.

Ditto for the textproto files from #3340. I suspect you could create a new CODEOWNERS section at the bottom called "Static analysis check configurations" or something like that.

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -0,0 +1,617 @@
+/** Lists of files which should be exepmted for the different script checks. */
+object ExemptionsList {
+  /** List of files which should be exempted for the test file check. */
+  val TEST_FILE_CHECK_EXEMPTIONS_LIST = listOf(
```
Please file an issue to add test suites with all of these listed as checkboxes. We can split that up into groups of starter issues later once we have a clear set of instructions for how to resolve these. Please also add a TODO here referencing that issue.

I suggest doing this after removing the scripts files from the list below.

Sparsh1212 replied:
Okay, will do.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,84 @@
+      val testFileCheck: TestFileCheck = TestFileCheck()
+
+      // a list of all the prod files present in the repo.
+      val prodFilesList: MutableList<String> = ArrayList()
```
```suggestion
      val prodFilesList = mutableListOf<String>()
```

More idiomatic. Ditto elsewhere.

Sparsh1212 replied:
Outdated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/TestFileCheck.kt:
```diff
@@ -0,0 +1,84 @@
+   * @return log the failures
+   */
+  private fun logFailures(matchedFileNames: List<String>) {
+    if (matchedFileNames.size != 0) {
```
Prefer ``isNotEmpty``

Sparsh1212 replied:
Done

BenHenning reviewed (changes requested):
Thanks @Sparsh1212. Took a first pass at this rather than a full pass since a lot of comments from the previous 2 PRs apply here, as well. I suggest addressing all the comments between the three PRs together since almost all of them apply to all three PRs, then send them each back to me.

Sparsh1212 committed:
Review suggestions part 1

Sparsh1212 committed:
Implement review suggestions part 2

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
resolve merge conflict

Sparsh1212 committed:
update static_checks

Sparsh1212 committed:
add test to repository file

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
Do review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 commented on scripts/src/java/org/oppia/android/scripts/common/ScriptExemptions.kt:
```diff
@@ -0,0 +1,621 @@
+    "GaeSubtopicSummary.kt",
+    "NetworkApiKey.kt",
+    "NetworkModule.kt",
+    "ScriptExemptions.kt",
+    "ScriptResultConstants.kt",
```
@BenHenning Added these two files because they only contain constants or a list of constants. So, I think it is fine to exempt from the check.

BenHenning replied:
The constants one came to mind to me as well--we may need to find a way to properly exempt files which do not ever need to have tests. Maybe if the file doesn't have an interface, annotation, function, or class?

Or, maybe we should just move the constants to their corresponding files instead of having a shared constants file. Does that file actually provide any specific benefits?

Sparsh1212 replied:
Will remove the constants file as it is not of any benefit.
Will do this after merging this PR with #3340.

Sparsh1212 replied:
Update: This is now removed.

Sparsh1212 commented:
@BenHenning 
The TestFile CI check is failing, because some files in `devoptions` and some other prod files were added without any test files. Please let me know what should we do about them.

```
No test file found for:
- /home/runner/work/oppia-android/oppia-android/testing/src/main/java/org/oppia/android/testing/DisableAccessibilityChecks.kt
- /home/runner/work/oppia-android/oppia-android/testing/src/main/java/org/oppia/android/testing/AccessibilityTestRule.kt
- /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsFragment.kt
- /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsFragmentPresenter.kt
- /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityPresenter.kt
- /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsStarter.kt
- /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsStarterImpl.kt
- /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsModule.kt
- /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsStarterModule.kt
```

Sparsh1212 committed:
Add Ben as a codeowner for the ScriptExemptions file

Sparsh1212 commented:
> Thanks @Sparsh1212. Took a first pass at this rather than a full pass since a lot of comments from the previous 2 PRs apply here, as well. I suggest addressing all the comments between the three PRs together since almost all of them apply to all three PRs, then send them each back to me.

Thanks, @BenHenning.
The PR is updated based on the earlier suggestions and on this one. PTAL.

Sparsh1212 committed:
bazel files nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Introduce a library for the regex assets

BenHenning commented:
@Sparsh1212 please assign back to me once this PR's chain (& subsequently this PR) has been updated to be based off of #3374.

Sparsh1212 committed:
Make directory structure same as that of #3374

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make directory structure similar to #3374

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
merge xml-syntax-checks

Sparsh1212 committed:
add new files to test file script exemptions and nit fixes

Sparsh1212 commented:
@BenHenning This PR's directory structure is now updated based on #3374.
PTAL.

Sparsh1212 commented:
> @BenHenning
> The TestFile CI check is failing, because some files in `devoptions` and some other prod files were added without any test files. Please let me know what should we do about them.
> 
> ```
> No test file found for:
> - /home/runner/work/oppia-android/oppia-android/testing/src/main/java/org/oppia/android/testing/DisableAccessibilityChecks.kt
> - /home/runner/work/oppia-android/oppia-android/testing/src/main/java/org/oppia/android/testing/AccessibilityTestRule.kt
> - /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsFragment.kt
> - /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsFragmentPresenter.kt
> - /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityPresenter.kt
> - /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsStarter.kt
> - /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsStarterImpl.kt
> - /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsModule.kt
> - /home/runner/work/oppia-android/oppia-android/app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsStarterModule.kt
> ```

Added to `TEST_FILE_CHECK_EXEMPTIONS_LIST` as discussed during the meet.

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
diable ktlint max-line-length

Sparsh1212 committed:
disable ktlint-max-line

Sparsh1212 committed:
disable ktlint max-length

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
remove testonly attribute from tests

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add new files to script exemptions

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

BenHenning commented on .github/CODEOWNERS:
```diff
@@ -228,3 +228,4 @@ WORKSPACE @BenHenning
 
 # Static analysis check configuration files ownership.
 /scripts/assets/ @BenHenning
+/scripts/src/java/org/oppia/android/scripts/common/ScriptExemptions.kt @BenHenning
```
Please add EOF newline.

Sparsh1212 replied:
Added.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/ScriptExemptions.kt:
```diff
@@ -0,0 +1,632 @@
+object ScriptExemptions {
+  /** List of files which should be exempted for the test file check. */
+  val TEST_FILE_CHECK_EXEMPTIONS_LIST = listOf(
+    "TestCoroutineDispatcherTestBase.kt",
```
Hmm. Should these have relative paths to the project root? I'm thinking about the case where two files happen to have the same name (it's technically possible).

Also, once we figure out the final format (i.e. relative paths) & layout (i.e. textproto per other comment), please lexicographically sort these so that they're easier to find (the paths will really help with that, I think).

Sparsh1212 replied:
Revamped to relative path approach, is also lexicographically sorted.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/ScriptExemptions.kt:
```diff
@@ -0,0 +1,632 @@
+ * have tests. If you think there's a valid exception, please specifically call this out in your PR
+ * description.
+ */
+object ScriptExemptions {
```
How hard would it be to move this to be a textproto, instead? Keeping all script configurations within textproto files seems kind of nice.

If we went that route, I suggest moving the 'NOTE TO DEVELOPERS' bit as output for the script itself. E.g. something like:

"File X does not have a corresponding test file. If this is correct, please update scripts/assets/script_exemptions.textproto. Note that, in general, all new files should have tests. If you choose to add an exemption, please specifically call this out in your PR description."

WDYT?

Sparsh1212 replied:
Sounds good to me. Revamped to the textproto approach. PTAL.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/ScriptExemptions.kt:
```diff
@@ -0,0 +1,632 @@
+    "DeveloperOptionsStarterImpl.kt",
+    "DeveloperOptionsFragmentPresenter.kt",
+    "DeveloperOptionsModule.kt",
+    "CommandResult.kt",
+    "CommandExecutor.kt",
```
Are these still needed?

Sparsh1212 replied:
Will remove them, once I take a merge from #3340. so that, the checks won't fali.

Sparsh1212 replied:
They are kotlin files and they don't have a corresponding TestFile present, so it looks like we need to add them to the exemption list.

Sparsh1212 replied:
Issue filed: https://github.com/oppia/oppia-android/issues/3436

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt:
```diff
@@ -0,0 +1,85 @@
+
+  searchFiles.forEach { file ->
+    if (file.name.endsWith("Test.kt")) {
+      testFilesList.add(file)
```
Ditto here (per a comment from an earlier PR): prefer creating the two lists here in a more Kotlin-idiomatic way.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt:
```diff
@@ -0,0 +1,85 @@
+ * @param prodFileName name of the prod file
+ * @return potential name of the test file
+ */
+private fun retrievePotentionalTestFileName(prodFileName: String): String {
```
Perhaps instead: ``computeExpectedTestFileName`` might be clearer? We actually expect such a file to exist.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt:
```diff
@@ -0,0 +1,85 @@
+ * @return potential name of the test file
+ */
+private fun retrievePotentionalTestFileName(prodFileName: String): String {
+  return "${prodFileName.removeSuffix(".kt")}Test.kt"
```
I think you could alternatively pass in the actual ``File`` and use ``File.nameWithoutExtension`` to simplify this a bit.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt:
```diff
@@ -0,0 +1,85 @@
+ */
+private fun logFailures(matchedFiles: List<File>) {
+  if (matchedFiles.isNotEmpty()) {
+    println("No test file found for:")
```
```suggestion
    println("No test file(s) found for:")
```

To fix grammar. I suggest double-checking your other scripts' outputs to verify the same issue doesn't exist for them.

Sparsh1212 replied:
Outdated.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/testfile/TestFileCheckTest.kt:
```diff
@@ -0,0 +1,94 @@
+import java.io.PrintStream
+
+/** Tests for [TestFileCheck]. */
+class TestFileCheckTest {
```
We ought to have at least 1 test for exemptions (though there are clearly too many to test them all, and we don't necessarily want to test all of them, anyway).

Sparsh1212 replied:
Added.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212. Left a few comments. PTAL.

Sparsh1212 committed:
Apply review suggestions

Sparsh1212 committed:
ktlint fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Apply nit fixes

Sparsh1212 committed:
Apply nit fixes

Sparsh1212 commented:
@BenHenning The PR is updated as per the nit suggestions. However, the text proto assets pattern is followed similar to that of the earlier approach of #3340. Once, we are finalized there, I will update this to reflect the final structure. PTAL.

Sparsh1212 committed:
Remove script constants

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
add tests for XML syntax error handler

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
add todo

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add todo

Sparsh1212 commented:
@BenHenning PTAL.

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

BenHenning commented on scripts/BUILD.bazel:
```diff
@@ -77,3 +78,22 @@ kt_jvm_binary(
+    test_file_exemptions_name = ["test_file_exemptions"],
+)
+
+kt_jvm_library(
```
This should also be marked as testonly. Ditto for earlier PRs if it applies.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/BUILD.bazel:
```diff
@@ -77,3 +78,22 @@ kt_jvm_binary(
+kt_jvm_library(
+    name = "test_file_check_assets",
+    data = TEST_FILE_EXEMPTION_ASSETS,
+    visibility = [":oppia_script_binary_visibility"],
```
Should this be restricted to test visibility since that's the only place where it's needed? Ditto for earlier PRs if applicable.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/RepositoryFile.kt:
```diff
@@ -40,7 +40,7 @@ class RepositoryFile() {
           file.isFile &&
           file.name.endsWith(expectedExtension) &&
-          file.name !in exemptionsList
+          retrieveRelativeFilePath(file, repoPath) !in exemptionsList
```
Maybe pull this into the PR where RepositoryFile was introduced for simplicity?

Sparsh1212 replied:
Done in the regex-checks PR.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/proto/BUILD.bazel:
```diff
@@ -32,3 +32,15 @@ java_lite_proto_library(
 )
+
+proto_library(
+    name = "script_exemptions_structure_proto",
```
Same as earlier PR: strike 'structure' since we generally just base the target name on the proto filename.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/proto/script_exemptions.proto:
```diff
@@ -0,0 +1,12 @@
+option java_package = "org.oppia.android.scripts.proto";
+option java_multiple_files = true;
+
+// Exemptions for the script checks.
```
This seems too vague--please add more details.

Sparsh1212 replied:
Added.

BenHenning replied:
As discussed in meeting, prefer to make these structures & their documentation specific to individual checks since it seems preferable to have distinct textprotos for each exemption list. Keeping the containing proto file generic seems fine.

Ditto for the field name (consider like 'exemptedFilePath' or similar).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/proto/script_exemptions.proto:
```diff
@@ -0,0 +1,12 @@
+
+// Exemptions for the script checks.
+message ScriptExemptions {
+  // List of all the exemptions.
+  repeated string exempt = 1;
```
What is an exemption? What's being exempted? How is it formatted? More details and a potential rename are needed here, I think.

Consider that this is the documentation for how to fill in the textproto file. Missing details makes that potentially much more difficult.

Sparsh1212 replied:
Added.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+"""
+Libraries corresponding to TestFile based check to ensure that all the production (all kotlin files)
```
```suggestion
Libraries corresponding to test file-related checks, such as ensuring that all the production (all kotlin files)
```

(This package documentation could be a bit more vague rather than overly specific to the current script since more scripts may be added in the future).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt:
```diff
@@ -0,0 +1,103 @@
+ */
+fun main(vararg args: String) {
+  // Path of the repo to be analyzed.
+  val repoPath = args[0] + "/"
```
I suggest using string interpolation here, too.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/testfile/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+"""
+Tests corresponding to TestFile based check to ensure that all the production (all kotlin files)
```
Reminder to update this per my earlier comment for the production package documentation for this package.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/testfile/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+    deps = [
+        "//scripts:test_file_check_assets",
+        "//scripts/src/java/org/oppia/android/scripts/testfile:test_file_check_lib",
+        "//testing:assertion_helpers",
```
Needed?

Sparsh1212 replied:
Yes, it's needed here, as we are throw error when the script fails, so we need to assert that in the tests.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt:
```diff
@@ -0,0 +1,103 @@
+ * @return proto class from the parsed textproto file
+ */
+private fun loadTestFileExemptionsProto(): ScriptExemptions {
+  val protoBinaryFile = File("scripts/assets/test_file_exemptions.pb")
```
Suggest extracting the path + filename - extension to a variable and using it both here & the error message above (in case it moves--both should stay in sync).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/testfile/TestFileCheckTest.kt:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    assertThat(exception).hasMessageThat().contains(TEST_FILE_CHECK_FAILED_OUTPUT_INDICATOR)
+    assertThat(outContent.toString()).contains("${retrieveTestFilesDirectoryPath()}/ProdFile2.kt")
```
Similar to earlier tests, we should actually validate that we're providing sufficient context & verifying the script's formulated error messages. Those are specifically designed to convey information in a certain way & it's important to continue that.

It's less important to verify the output of downstream libraries that produce errors for us (such as for the XML check), but the errors we output should be validated.

Ditto for earlier PRs if applicable.

Sparsh1212 replied:
Apologies! This is now rectified and updated. PTAL.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/testfile/TestFileCheckTest.kt:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  @Test
+  fun testTestFileCheck_prodFileWithCorrespondingTestFile_testFileIsPresent() {
```
It's a bit pathalogical, but maybe add a test for an empty directory.

Sparsh1212 replied:
Added.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212! Had a few more comments. I've reviewed everything except the textproto file--will do that after this latest round is addressed.

A few of my comments may affect prior PRs, too. Overall, I think the script is looking really good. Most of my comments are nits at this point.

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
add test

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add testOnly

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 commented:
> Thanks @Sparsh1212! Had a few more comments. I've reviewed everything except the textproto file--will do that after this latest round is addressed.
> 
> A few of my comments may affect prior PRs, too. Overall, I think the script is looking really good. Most of my comments are nits at this point.

Thanks @BenHenning everything addressed. PTAL.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/BUILD.bazel:
```diff
@@ -1,6 +1,6 @@
-Libraries corresponding to TestFile based check to ensure that all the production (all kotlin files)
-files have a corresponding test file present.
+Libraries corresponding to test file-related checks, such as ensuring that all the production
+(all kotlin files) files have a corresponding test file present.
```
Please capitalize Kotlin here & elsewhere.

Sparsh1212 replied:
Done.

BenHenning reviewed (commented):
Thanks @Sparsh1212. Leaving a few follow-ups per our meeting.

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
fix textproto file

Sparsh1212 committed:
merge-regex-checks

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

BenHenning commented on scripts/src/java/org/oppia/android/scripts/proto/script_exemptions.proto:
```diff
@@ -5,18 +5,17 @@ package proto;
-  // add 'exemption: "app/src/main/java/org/oppia/android/app/activity/ActivityComponent.kt"' at
-  // the correct lexicographical position of the 'test_file_exemptions.textproto' file.
-  repeated string exemption = 1;
+// Exemptions for the test-file check.
+// Here, exemption implies those files, which we do not wish to check for the test-file presence.
```
Nit: combine these into one doc comment rather than having the extra line between them.

Though, please do wrap at the 100 character limit like normal.

Also, nit on phrasing:

```
Exemptions indicate files for which we should not check for having test files.
```

(Slightly reworded for clarity).

Sparsh1212 replied:
Done.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212. Generally LGTM, but two questions:

1. Can you please bring this branch up-to-date with develop now that #3341 is merged & update the title to no longer indicate it's blocked
2. Can you clear the textproto file & show what the CI failure looks like for an empty exemption list? I'm interested in verifying the failure cases in a real CI scenario
3. After (2), please restore the textproto file so that CI is green (the result should be 1 commit with the specific failures requested from (2) and the latest commit passing)

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 commented:
@BenHenning 
The test commit added!
Also, note that one test will fail, because we have also removed the exempted file which is needed by the exempted file test. So, its a expected behaviour.

Sparsh1212 committed:
add exemptions to textproto

Sparsh1212 committed:
sort failures lexicographically

BenHenning reviewed (approved):
Thanks @Sparsh1212. LGTM! Just waiting for green CI to merge.

BenHenning commented:
@Sparsh1212 looks like the new check is failing in CI--please fix.

seanlip commented:
I'm a bit confused ... why is Oppiabot assigning me to this PR? I don't code-own any files in it.

Sparsh1212 committed:
Merge branch 'develop' into test-file-check

Sparsh1212 committed:
add new files to exemption list

BenHenning commented:
@seanlip looks like a potential bug. You were auto-added as a reviewer because a base PR changed the codeowners file. I guess Oppiabot assumes that reviewers always get auto-assigned even if they aren't a codeowner or at-mentioned?

BenHenning commented:
Thanks @Sparsh1212. Everything is looking good. Enabling auto-merge.

seanlip commented:
Oh, weird. Hm, I'm not too bothered about it, but if it recurs it might be worth fixing on Oppiabot's end. /cc @jameesjohn for info.

jameesjohn commented:
> @seanlip looks like a potential bug. You were auto-added as a reviewer because a base PR changed the codeowners file. I guess Oppiabot assumes that reviewers always get auto-assigned even if they aren't a codeowner or at-mentioned?

Yes. oppiabot assumes that reviewers who are yet to review a PR are to be assigned.
